### ISSUE TITLE
HPFX falls behind (but DD works fine)

### ISSUE THREAD

**Issue by endlisnis** [MetPX/sarracenia#1331]
I've been trying to switch from `dd` to `hpfx`, but I keep running into unexpected problems.

For the last 2 days, whenever I start listening to HPFX, my client slowly falls behind and never catches up.  I try this exact same config on dd, and it works fine.

Here's my config:
```
broker amqps://hpfx.collab.science.gc.ca

queue_name q_anonymous.${HOSTNAME}.20241013
instances 4

directory ${PWD}/swobMirror/
# All stations
strip 3
topicPrefix v02.post
subtopic *.WXO-DD.observations.swob-ml.#
batch 64

mirror True
accept .*
expire 4h
acceptSizeWrong off
inflight tmp/

rxq_name ${PWD}/rxpipe
flowCallback rxqueue_gzip.RxQueue_gzip
```

And I've attached my log from one of the instances today.

It has a lot of "too busy to retry" messages.  I'm not sure if that's relevant or not.

I also get a lot of "ack failed" messages.

```
$ sr3 --version
3.00.56
```

[debug.log.gz](https://github.com/user-attachments/files/18041039/debug.log.gz)

**Comment by endlisnis** [comment]
I tried restarting it, but it didn't improve.

```
$ sr3 status subscribe/hpfx
status: 
Component/Config     Processes   Connection        Lag                              Rates                                        
                     State   Run Retry  msg data   Que   LagMax   LagAvg     Last  %rej     pubsub   messages     RxData     TxData 
                     -----   --- -----  --- ----   ---   ------   ------     ----  ----     ------   --------     ------     ------ 
subscribe/hpfx       lag     4/4     0 100%  17%  3838   13m53s    10m2s    5.55s  3.8%     357B/s       2m/s  19.2KiB/s       0B/s 
      Total Running Configs:   1 ( Processes: 4 missing: 0 stray: 0 )
                     Memory: uss:156.6MiB rss:249.0MiB vms:417.0MiB 
                   CPU Time: User:29.29s System:5.36s 
           Pub/Sub Received: 7m/s (1.2KiB/s), Sent:  0m/s (0B/s) Queued: 15352 Retry: 0, Mean lag: 607.55s
              Data Received: 7f/s (65.9KiB/s), Sent: 0f/s (0B/s) 
```

**Comment by endlisnis** [comment]
CPU utilization is normal (below 10% per instance).

**Comment by petersilva** [comment]
the too busy message is usually bogus... will show up a lot less often after next release (fix in dev branch: https://github.com/MetPX/sarracenia/pull/1292 )

Maybe try more instances?  I don't know any reason why it wouldn't keep up...  disk i/o is sort of local on dd (SAN) but kind of remote on hpfx (NAS.) maybe hpfx disk is slower...

**Comment by endlisnis** [comment]
OK.  When I increase the number of instances to 12, it can keep up.  That seems kind sad that HPFX is that much slower.  But I suppose this is not a bug in sr3.

**Comment by petersilva** [comment]
Some github projects get their shorts in a knicker because people ask questions by opening issues... We're fine with it.  We just add tags to classify them appropriately.  Also we can take that feeback on board and perhaps see if we can speed up hpfx.

your feedback is really appreciated.  It's an interesting observation, and we were not aware of it.

**Comment by endlisnis** [comment]
I'm not sure what changed, but I accidentally started using the HPFX server again today and 4 instances was able to keep up.  I'm using the same setup as before.

**Comment by endlisnis** [comment]
And I just noticed that my HPFX client started falling behind last week, after months of doing just fine with 4 threads.  DD works perfectly with 4 threads.

**Comment by reidsunderland** [comment]
Thanks for letting us know. We've noticed that HPFX gets a bit overwhelmed at peak times and are looking into performance improvements, but don't have a timeline quite yet. 

There was a small change on June 19 that we hoped would improve the situation, but it didn't seem to help. Is June 18 when you noticed the performance got worse or did it start before then?

**Comment by endlisnis** [comment]
This is not just a peak times.  It's literally 24/7.

It had been working fine for several months, but then I noticed a problem about 10 days ago.  It was not June 18th.

As I described earlier, when I do independent testing, the HPFX server seems fine; but when sr3 uses it, it can't keep up.

**Comment by reidsunderland** [comment]
Interesting. Just to clarify: "independent testing" - does that mean just downloading the data from the server without using any AMQP messages/notifications?

**Comment by endlisnis** [comment]
Yes, by "independent testing", I mean timing downloading a single file, or using a single session to download a few dozens files.  But all of it with just curl or python -- not using AMQP and not using `sr3`.

And iterating over each of those a few dozen times to get an average.  In my tests like that, HPFX shows better results than dd.

**Comment by matthewdarwin** [comment]
FWIW, I always use my own script to download files.  Sometimes that also falls behind on HPFX (I get alert telling me I have several thousand files pending download).   Either that or something is broken posting the files to HPFX server such there is sometimes a backload to download.  (I'm downloading all the citypage and swob files, so its a lot of files)

**Comment by reidsunderland** [comment]
This is helpful information, it seems to point to a broker performance problem in addition to the disk I/O limitations. I made a small tweak yesterday (2025-06-23) morning that might have helped, although I don't see any changes in our monitoring.
